src/io/hdf_stl.h

```diff
@@ -123,8 +123,6 @@ template<> struct h5data_proxy<std::string>
       }
       ref.resize(dim_out);
       herr_t ret = H5Dread(dataset, datatype, H5S_ALL, H5S_ALL, xfer_plist,&(ref[0]));
-      // Erase trailing null character
-      ref.erase (dim_out-1, 1);
```

Comment:
Unfortunately I can't find my old case, it happened on a KNL box.
At that time, a string search failed because of the missing last character. I commented the trailing line in the h5 interface but I noticed the last character was not the same as h5. So I just extend the original string written to the h5 to make the string searching work.
Anyway this change is save.
Mark's fix the definitely correct. And the tests are useful to diagnose problem to identify the particular machine.
OK to merge.